src/gridfs/download.rs

```diff
@@ -0,0 +1,181 @@
+    {
+        self.download_to_tokio_writer(id, &mut destination.compat_write())
+            .await
+    }
+
+    /// Downloads the contents of the stored file specified by `filename` and writes the contents to
+    /// the `destination`. If there are multiple files with the same filename, the `revision` in the
+    /// options provided is used to determine which one to download. If no `revision` is specified,
+    /// the most recent file with the given filename is chosen.
+    pub async fn download_to_tokio_writer_by_name<T>(
+        &self,
+        filename: impl AsRef<str>,
+        destination: &mut T,
+        options: impl Into<Option<GridFsDownloadByNameOptions>>,
+    ) -> Result<()>
+    where
+        T: tokio::io::AsyncWrite + std::marker::Unpin,
+    {
+        let revision = options.into().and_then(|opts| opts.revision).unwrap_or(-1);
+        let (sort, skip) = if revision >= 0 {
+            (1, revision)
+        } else {
+            (-1, -revision - 1)
+        };
+        let options = FindOneOptions::builder()
+            .sort(doc! { "uploadDate": sort })
+            .skip(skip as u64)
+            .read_concern(self.read_concern().cloned())
+            .selection_criteria(self.selection_criteria().cloned())
+            .build();
+
+        let file = match self
+            .inner
+            .files
+            .find_one(doc! { "filename": filename.as_ref() }, options)
+            .await?
+        {
+            Some(fcd) => fcd,
+            None => {
+                return Err(ErrorKind::InvalidArgument {
+                    message: format!("couldn't find file with name {}", &filename.as_ref()),
+                }
+                .into());
+            }
+        };
+
+        self.download_to_tokio_writer_common(file, destination)
+            .await
+    }
+
+    /// Downloads the contents of the stored file specified by `filename` and writes the contents to
+    /// the `destination`. If there are multiple files with the same filename, the `revision` in the
+    /// options provided is used to determine which one to download. If no `revision` is specified,
+    /// the most recent file with the given filename is chosen.
+    pub async fn download_to_futures_0_3_writer_by_name<T>(
+        &self,
+        filename: impl AsRef<str>,
+        destination: &mut T,
+        options: impl Into<Option<GridFsDownloadByNameOptions>>,
+    ) -> Result<()>
+    where
+        T: futures_util::io::AsyncWrite + std::marker::Unpin,
+    {
+        self.download_to_tokio_writer_by_name(filename, &mut destination.compat_write(), options)
+            .await
+    }
+
+    async fn download_to_tokio_writer_common<T>(
+        &self,
+        file: FilesCollectionDocument,
+        destination: &mut T,
+    ) -> Result<()>
+    where
+        T: tokio::io::AsyncWrite + std::marker::Unpin,
+    {
+        let total_bytes_expected = file.length;
+        let chunk_size = file.chunk_size as u64;
```

Comment:
Since this value is coming from the DB, we may want to validate that it is within the range of `u64`, returning an error if it isn't.

Reply:
Yeah, the concern is that the data gets messed up in some other way by a different driver or application (i.e. it's untrusted). Basically, don't want to accidentally use a junk value in the logic that follows. Seems mostly harmless, but I think it would be good to explicitly verify that.

Reply:
Ah ok my bad, I thought it was an `i32` or `i64` being cast to a `u64`.